adapter/src/test/kotlin/org/javacs/ktda/GlobalWorkspaceTest.kt

```diff
@@ -0,0 +1,242 @@
+import io.mockk.verify
+import org.eclipse.lsp4j.debug.*
+import java.nio.file.Path
+import java.nio.file.Paths
+import org.eclipse.lsp4j.debug.services.IDebugProtocolClient
+import org.hamcrest.Matchers
+import org.javacs.ktda.adapter.KotlinDebugAdapter
+import org.javacs.ktda.jdi.launch.JDILauncher
+import org.junit.After
+import org.junit.Assert.assertThat
+import org.junit.Before
+import org.hamcrest.Matchers.equalTo
+import org.junit.Test
+import java.lang.Thread
+import java.util.concurrent.CountDownLatch
+
+/**
+ * Defines the basic debugging scenarios
+ * that should be tested on all supported sample applications.
+ */
+abstract class GlobalWorkspaceTest(
+    relativeWorkspaceRoot: String,
+    private val mainClass: String,
+    private val defaultVmArguments: String = "",
+    private val buildCommand: Pair<String, String>
+) {
+    val absoluteWorkspaceRoot: Path =
+        Paths.get(GlobalWorkspaceTest::class.java.getResource("/Anchor.txt").toURI()).parent.resolve(
+            relativeWorkspaceRoot
+        )
+    var client = mockk<IDebugProtocolClient>(relaxed = true)
+    lateinit var debugAdapter: KotlinDebugAdapter
+    val latch = CountDownLatch(1)
+
+    @Before
+    fun startDebugAdapter() {
+        // Reset the mock debug protocol client
+        client = mockk(relaxed = true)
+
+        // Build the project first
+        val process = ProcessBuilder(buildCommand.first, buildCommand.second)
+            .directory(absoluteWorkspaceRoot.toFile())
+            .inheritIO()
+            .start()
+        process.waitFor()
+        assertThat(process.exitValue(), equalTo(0))
+
+        debugAdapter = JDILauncher()
+            .let(::KotlinDebugAdapter)
+            .also {
+                it.connect(client)
+                val configDone = it.configurationDone(ConfigurationDoneArguments())
+                it.initialize(InitializeRequestArguments().apply {
+                    adapterID = "test-debug-adapter"
+                    linesStartAt1 = true
+                    columnsStartAt1 = true
+                }).join()
+                // Slightly hacky workaround to ensure someone is
+                // waiting on the ConfigurationDoneResponse. See
+                // KotlinDebugAdapter.kt:performInitialization for
+                // details.
+                Thread {
+                    configDone.join()
+                }.start()
+                // Wait until the thread has blocked on the future
+                while (configDone.numberOfDependents == 0) {
+                    Thread.sleep(100)
+                }
+            }
+
+        // Print output on commandline for test debugging
+        slot<OutputEventArguments>().also { outputEvent ->
+            every { client.output(capture(outputEvent)) }.answers { println(outputEvent.captured.output) }
+        }
+        // And release the latch when the debug session is terminated/Done
+        every { client.terminated(any()) }.answers { latch.countDown() }
```

Comment:
The latch now only opens when the debugging session is ended.

Every tests can configure how the client behaves, wait for the session to be done, and then verify the mock to see if the server behaved as expected.